contracts/libs/MarginAccount.sol

```diff
@@ -1,12 +1,220 @@
+    }
+
+    /**
+     * @dev increase the collateral balance in a vault
+     * @param _vault The vault to which the protocol is adding the collateral
+     * @param _collateralAsset The address of the _collateralAsset which the protocol is adding to the user's vault
+     * @param _amount The number of _collateralAsset the protocol is adding to the user's collateral position in the user's vault
+     * @param _index The index of _collateralAsset in the user's vault.collateralAssets array
+     */
+    function _addCollateral(
+        Vault storage _vault,
+        address _collateralAsset,
+        uint256 _amount,
+        uint256 _index
+    ) internal {
+        // Valid indexes in any array are between 0 and array.length - 1.
+        // If adding amount into an existing long otoken, make sure _index in the range of 0->length-1
+        if ((_index >= _vault.collateralAssets.length) && ((_index >= _vault.collateralAmounts.length))) {
+            _vault.collateralAssets.push(_collateralAsset);
+            _vault.collateralAmounts.push(_amount);
+        } else {
+            require(
+                (_vault.collateralAssets[_index] == _collateralAsset) ||
+                    (_vault.collateralAssets[_index] == address(0)),
+                "MarginAccount: invalid collateral token position"
+            );
+
+            _vault.collateralAmounts[_index] = _vault.collateralAmounts[_index].add(_amount);
+            _vault.collateralAssets[_index] = _collateralAsset;
+        }
+    }
+
+    /**
+     * @dev decrease the collateral balance in a vault
+     * @param _vault The vault from which the protocol is removing the collateral
+     * @param _collateralAsset The address of the _collateralAsset which the protocol is removing from the user's vault
+     * @param _amount The number of _collateralAsset the protocol is removing from the user's collateral position in the user's vault
+     * @param _index The index of _collateralAsset in the user's vault.collateralAssets array
+     */
+    function _removeCollateral(
+        Vault storage _vault,
+        address _collateralAsset,
+        uint256 _amount,
+        uint256 _index
+    ) internal {
+        //Check the token is the same as vault.collateral[idx]
+        require(
+            ((_index < _vault.collateralAssets.length) && (_vault.collateralAssets[_index] == _collateralAsset)),
+            "MarginAccount: collateral token address mismatch"
+        );
+
+        _vault.collateralAmounts[_index] = _vault.collateralAmounts[_index].sub(_amount);
+    }
 
+    /**
+     * @dev remove everything in a vault. Reset short, long and collateral assets and amounts arrays to [].
+     * @param _vault The vault that the user is clearing.
+     */
+    function _clearVault(Vault storage _vault) internal {
+        delete _vault.shortAmounts;
+        delete _vault.longAmounts;
+        delete _vault.collateralAmounts;
+        //TODO: are the following required?
```

Comment:
TODO: make a decision on this

Reply:
yes @antoncoding @haythem96 lmk what you think bout htis. i know haythem you had a completely different proposal entirely

Reply:
I think using delete is clean and it payback a little bit gas

Reply:
We should discuss this in today eng sync, I don't see why we need to clear vault and re-use them, def not gas efficient.

Reply:
we decided this right? resolving it